### PR TITLE
Fix bug with the "origin" parameter

### PR DESCRIPTION
**Pull request by Klortho**
This fixes a bug in the "origin" parameter. I have a repo that has two remotes:
- origin - points to an internal repo hosting service
- github - GitHub, obviously.

So I set the "origin: 'github'", but it failed, because it was trying to push from the temporary ".publish" repo, looking for a remote there, called "github". But the remote in that temporary repo is always called "origin".

It turned out that the fix was as simple as changing `origin -> "origin"` in two places, but I spent a really long time figuring it out, and trying to write a test for it. You'll see that I rewrote your test.js file pretty extensively -- that's mainly because I had a really hard time getting it to work. It was relying on a github api token, and using a real github repo for the tests. I changed it to use a "file:" repo for the tests -- so they should be much more reliable, since no external dependencies.

In the process of figuring out how the tests worked, I reformatted it some for readability (shorter line lengths), since I am working on a small-screen laptop, and like I said, I've been staring at it for hours. I hope you don't mind the formatting changes.

### CODE REVIEW

**Comment by shinnn** [comment]
Has a lot of lint errors. https://travis-ci.org/shinnn/gulp-gh-pages/builds/104067005#L478-L478

**Comment by Klortho** [comment]
Okay, everything's passing now. I took the liberty of adding an eslint rule for line length maximum at 90 characters. Coverage went down slightly, but I'd note that I *added* a test for a permutation of input parameters that wasn't tested before.

I hope you either accept this pull request, or else remove the "origin" configuration parameter completely, because it is badly broken, and it could save the next guy who tries to use it a lot of time.

Thanks for such a useful tool!

**Comment by shinnn** [comment]
> I took the liberty of adding an eslint rule for line length maximum at 90 characters.

No thanks.

**Comment by Klortho** [comment]
> No thanks.

Then go ahead and delete it. You're the one who responded to my lengthy description of a problem and solution with "Has a lot of lint errors". One major source of lint errors were trailing spaces on lines -- seriously?

**Comment by shinnn** [comment]
> Then go ahead and delete it.

No. Should be deleted in this pull request.

**Comment by Klortho** [comment]
done

**Comment by shinnn** [comment]
@Klortho Thanks. I'd like you to separate test the refactoring part from this pull request. Can you create another one? I don't care whether this pull request passes the test or not (except for linting errors).

**Comment by shinnn** [comment]
> I hope you either accept this pull request, or else remove the "origin" configuration parameter completely, because it is badly broken, and it could save the next guy who tries to use it a lot of time.

After seeing this pull request,

1. I decided to remove `origin` option that has been confusing a lot of users.
2. Before removing `origin` option, I'll publish one patch-level update to npm for compatibility.

**Comment by Klortho** [comment]
> I decided to remove origin option that has been confusing a lot of users.

I suspect it might have been confusing a lot of users because it's been broken. Note that this PR includes some updates to your README, the I think would make it's purpose more clear.

>  I'd like you to separate the test refactoring part from this pull request. Can you create another one? 

The test refactoring is a major improvement to this code -- you really shouldn't have dependencies on github, and (even worse) specific repositories in GitHub.  I'm not willing to spend more time on this -- I already spent way too much as it is. I had hoped you'd be grateful to get a big improvement to your module, for free.

File: README.md [hunk 1]
```diff
@@ -37,6 +37,16 @@ Now, you should be able to call your task by doing:
 gulp deploy
 ```
 
+If you want to push gh-pages to a remote other than "origin":
```
Comment (shinnn) [inline README.md]: Move it above the `Now, you should be able to call ...`.

**Comment by shinnn** [comment]
>  I'm not willing to spend more time on this

OK, so I will not directly merge this and cherrypick the commits instead.

File: test.js [hunk 2]
```diff
@@ -31,46 +31,26 @@ var files = [
   }),
   tmpFile
 ];
-var tmpRepoName = 'shinnn/css-wide-keywords';
 var remoteUrl;
 
-var accessToken = process.env.GH_ACCESS_TOKEN;
-var client;
+function unzip_test_repo(dest, done) {
```
Comment (shinnn) [inline test.js]: `unzipTestRepo`

File: test.js [hunk 3]
```diff
@@ -31,46 +31,26 @@ var files = [
...
+  var test_repo = 'test-git-repo.git';
+  var dest_path = path.join(__dirname, dest);
+  rimraf(dest_path, function() {
+    (new Zip(path.join(__dirname, test_repo + '.zip')))
+      .extractAllTo('.', true);
+    fs.renameSync(test_repo, dest);
```
Comment (shinnn) [inline test.js]: Use asynchronous `fs.rename` here.

File: test.js [hunk 4]
```diff
@@ -231,91 +195,126 @@ describe('gulp-gh-pages', function() {
...
+        process.chdir(path.resolve(__dirname, '..'));
+        done();
+      })
+      .on('end', function() {
+        done(new Error('Expected an error.'));
+      })
+      .end(tmpFile);
+    });
 });
+
```
Comment (shinnn) [inline test.js]: Not needed.

**Comment by shinnn** [comment]
Also, squash those four commits into two:

1. fix the bug with the origin parameter
2. get tests working without using GitHub

File: test.js [hunk 5]
```diff
@@ -7,12 +7,12 @@ var File = require('vinyl');
 var fs = require('graceful-fs');
 var ghPages = require('./');
 var git = require('./lib/git');
-var github = require('octonode');
-var logSymbols = require('log-symbols');
```
Comment (shinnn) [inline test.js]: Should also remove `log-symbols` from package.json.

File: test.js (hunk 4 above)
Comment (Klortho) [inline test.js]: What is not needed?

Comment (shinnn) [inline test.js]: Extra linebreak in L320.

**Comment by Klortho** [comment]
all done

File: README.md [hunk 6]
```diff
@@ -45,54 +54,54 @@ var ghPages = require('gulp-gh-pages');
 
 ### ghPages([*options*])
 
-*options*: `Object`
```
Comment (shinnn) [inline README.md]: Don't remove this two spaces.

File: README.md [hunk 7]
```diff
@@ -45,54 +54,54 @@ var ghPages = require('gulp-gh-pages');
 
 ### ghPages([*options*])
 
-*options*: `Object`  
+*options*: `Object`
 Return: `Object` ([stream.Transform](https://nodejs.org/api/stream.html#stream_class_stream_transform_1))
 
 #### options.remoteUrl
 
-Type: `String`
```
Comment (shinnn) [inline README.md]: Don't remove this two spaces.

File: README.md [hunk 8]
```diff
@@ -45,54 +54,54 @@ var ghPages = require('gulp-gh-pages');
...
-Type: `String`  
+Type: `String`
 Default: URL for the remote of the current dir (assumes a git repository)
 
-By default `gulp-gh-pages` assumes the current working directory is a git repository and uses its remote url. If your `gulpfile.js` is not in a git repository, or if you want to push to a different remote url, you can specify it. Ensure you have write access to the repository.
+By default `gulp-gh-pages` assumes the current working directory is a git repository and uses the URL of the remote designated by `origin`. If your `gulpfile.js` is not in a git repository, or if you want to push to a different remote url, you can specify it. Ensure you have write access to the repository.
 
 #### options.origin
 
-Type: `String`
```
Comment (shinnn) [inline README.md]: Don't remove this two spaces.

File: README.md [hunk 9]
```diff
@@ -45,54 +54,54 @@ var ghPages = require('gulp-gh-pages');
...
-Type: `String`  
+Type: `String`
 Default: `"origin"`
 
-Git remote.
+The Git remote to push the changes to. If `remoteUrl` is given, this is ignored.
 
 #### options.branch
 
-Type: `String`
```
Comment (shinnn) [inline README.md]: Don't remove this two spaces.

File: README.md [hunk 10]
```diff
@@ -45,54 +54,54 @@ var ghPages = require('gulp-gh-pages');
...
 
-Type: `String`  
+Type: `String`
 Default: `"gh-pages"`
 
 The branch where deploy will by done. Change to "master" for `username.github.io` projects.
 
 #### options.cacheDir
 
-Type: `String`
```
Comment (shinnn) [inline README.md]: Don't remove this two spaces.

File: README.md [hunk 11]
```diff
@@ -45,54 +54,54 @@ var ghPages = require('gulp-gh-pages');
...
 
-Type: `String`  
+Type: `String`
 Default: `.publish`
 
 Set the directory path to keep a cache of the repository. If it doesn't exist, gulp-gh-pages automatically create it.
 
 #### options.push
 
-Type: `Boolean`
```
Comment (shinnn) [inline README.md]: Don't remove this two spaces.

File: README.md [hunk 12]
```diff
@@ -45,54 +54,54 @@ var ghPages = require('gulp-gh-pages');
...
 
-Type: `Boolean`  
+Type: `Boolean`
 Default: `true`
 
 Allow you to make a build on the defined branch without pushing it to master. Useful for dry  run.
 
 #### options.force
 
-Type: `Boolean`
```
Comment (shinnn) [inline README.md]: Don't remove this two spaces.

File: README.md [hunk 13]
```diff
@@ -45,54 +54,54 @@ var ghPages = require('gulp-gh-pages');
...
 
-Type: `Boolean`  
+Type: `Boolean`
 Default: `false`
 
 Force adding files to the `gh-pages` branch, even if they are ignored by `.gitignore` or `.gitignore_global`.
 
 #### options.message
 
-Type: `String`
```
Comment (shinnn) [inline README.md]: Don't remove this two spaces.

File: test.js [hunk 14]
```diff
@@ -31,46 +31,25 @@ var files = [
   }),
   tmpFile
 ];
-var tmpRepoName = 'shinnn/css-wide-keywords';
 var remoteUrl;
```
Comment (shinnn) [inline test.js]: Define `remoteUrl` here.

```js
remoteUrl = 'file://' + path.join(__dirname, 'test-remote');
```

File: test.js [hunk 15]
```diff
@@ -78,21 +57,23 @@ describe('git operations on a repo', function() {
...
-    })
-    .then(function() {
-      done(new Error('Expected an error.'));
-    }, function(err) {
-      var expectedMsg = 'error: pathspec \'' +
-                        'non-existent-branch' +
-                        '\' did not match any file(s) known to git.\n';
-      assert.notEqual(err.message.indexOf(expectedMsg), -1);
-      done();
+  it('should throw an error when checking out a non existent branch',
```
Comment (shinnn) [inline test.js]:
```js
it('should throw an error when checking out a non existent branch', function(done) {
```

File: test.js [hunk 16]
```diff
@@ -110,91 +91,81 @@ describe('git operations on special repositories', function() {
...
-    })
-    .on('error', done)
-    .on('data', function(file) {
-      assert(file.isBuffer());
-    })
-    .on('end', function() {
-      fs.readFile(path.join(tmpDir, tmpStr), function(err, buf) {
-        assert.strictEqual(err, null);
-        assert.equal(String(buf), tmpStr);
+  it('should ignore empty vinyl file objects',
```
Comment (shinnn) [inline test.js]: Same as https://github.com/shinnn/gulp-gh-pages/pull/83/files#r50623117

File: test.js [hunk 17]
```diff
@@ -110,91 +91,81 @@ describe('git operations on special repositories', function() {
...
+        assert(file.isNull());
         done();
-      });
+      })
+      .end(new File());
     });
 
-    files.forEach(function(file) {
-      stream.write(file);
+  it('should emit an error when it takes files with stream contents',
```
Comment (shinnn) [inline test.js]: Same as https://github.com/shinnn/gulp-gh-pages/pull/83/files#r50623117

File: test.js [hunk 18]
```diff
@@ -110,91 +91,81 @@ describe('git operations on special repositories', function() {
...
+      .on('error', function(err) {
+        assert.equal(err.message, 'Stream content is not supported');
+        done();
+      })
+      .end(new File({contents: new PassThrough({objectMode: true})}));
     });
 
-    stream.end();
-  });
+  it('should push commits to an existing gh-pages branch',
```
Comment (shinnn) [inline test.js]: Same as https://github.com/shinnn/gulp-gh-pages/pull/83/files#r50623117

File: test.js [hunk 19]
```diff
@@ -110,91 +91,81 @@ describe('git operations on special repositories', function() {
...
+      });
 
-    files.forEach(function(file) {
-      stream.write(file);
+      stream.end();
     });
 
-    stream.end();
-  });
+  it('should not push any commits when no files has been changed',
```
Comment (shinnn) [inline test.js]: Same as https://github.com/shinnn/gulp-gh-pages/pull/83/files#r50623117

File: test.js [hunk 20]
```diff
@@ -110,91 +91,81 @@ describe('git operations on special repositories', function() {
...
-        return;
-      }
+      files.forEach(function(file) {
+        stream.write(file);
+      });
+
+      stream.end();
+    });
 
+  it('should create and checkout a branch',
```
Comment (shinnn) [inline test.js]: Same as https://github.com/shinnn/gulp-gh-pages/pull/83/files#r50623117

File: test.js [hunk 21]
```diff
@@ -231,91 +194,125 @@ describe('gulp-gh-pages', function() {
...
-      assert(file.isBuffer());
-    })
-    .on('end', function() {
-      readRemoveFile(path.join('__cache__', tmpStr), function(readErr, buf) {
-        assert.strictEqual(readErr, null);
-        assert.equal(String(buf), tmpStr);
-        done();
+  // To test this, we have to first create a dummy repo, and
+  // add a new remote to it
+  it('should work with a different remote for origin',
```
Comment (shinnn) [inline test.js]: Same as https://github.com/shinnn/gulp-gh-pages/pull/83/files#r50623117

File: test.js [hunk 22]
```diff
@@ -231,91 +194,125 @@ describe('gulp-gh-pages', function() {
...
-    .on('error', function(err) {
-      assert(err.code);
-      done();
-    })
-    .on('end', function() {
-      done(new Error('Expected an error.'));
-    })
-    .end(tmpFile);
-  });
+  it('should specify the cache directory path with `cacheDir` option',
```
Comment (shinnn) [inline test.js]: Same as https://github.com/shinnn/gulp-gh-pages/pull/83/files#r50623117

File: test.js [hunk 23]
```diff
@@ -231,91 +194,125 @@ describe('gulp-gh-pages', function() {
...
-      assert(err);
-      process.chdir(path.resolve(__dirname, '..'));
-      done();
-    })
-    .on('end', function() {
-      done(new Error('Expected an error.'));
-    })
-    .end(tmpFile);
-  });
+  it('should emit an error when the current directory is not a git repository',
```
Comment (shinnn) [inline test.js]: Same as https://github.com/shinnn/gulp-gh-pages/pull/83/files#r50623117

File: test.js [hunk 24]
```diff
@@ -231,91 +194,125 @@ describe('gulp-gh-pages', function() {
...
+      })
+      .on('end', function() {
+        done(new Error('Expected an error.'));
+      })
+      .end(tmpFile);
+    });
 
-  it('should emit an error when the current directory is not a git repository', function(done) {
-    process.chdir('/');
+  it('should emit an error when the remote URL is not a git repository\'s URL',
```
Comment (shinnn) [inline test.js]: Same as https://github.com/shinnn/gulp-gh-pages/pull/83/files#r50623117

File: test.js [hunk 25]
```diff
@@ -231,91 +194,125 @@ describe('gulp-gh-pages', function() {
...
-    .on('error', function(err) {
-      assert(/'https:\/\/example\.org\/' not found/.test(err.message));
-      done();
-    })
-    .on('end', function() {
-      done(new Error('Expected an error.'));
-    })
-    .end(tmpFile);
-  });
+  it('should emit an error when the user has no permission to push the repo',
```
Comment (shinnn) [inline test.js]: Same as https://github.com/shinnn/gulp-gh-pages/pull/83/files#r50623117

File: test.js [hunk 26]
```diff
@@ -231,91 +194,125 @@ describe('gulp-gh-pages', function() {
...
-    .on('error', function(err) {
-      assert(err);
-      done();
-    })
-    .on('end', function() {
-      done(new Error('Expected an error.'));
-    })
-    .end(tmpFile);
-  });
+  it('should emit an error when it fails to create a cache directory',
```
Comment (shinnn) [inline test.js]: Same as https://github.com/shinnn/gulp-gh-pages/pull/83/files#r50623117

File: test.js [hunk 27]
```diff
@@ -7,12 +7,12 @@ var File = require('vinyl');
 var fs = require('graceful-fs');
 var ghPages = require('./');
 var git = require('./lib/git');
-var github = require('octonode');
-var logSymbols = require('log-symbols');
+var gift = require('gift');
```
Comment (shinnn) [inline test.js]: Can you move this line right under `var ghPages`? I want to keep these `var` declarations in alphabetical order.

File: test.js [hunk 28]
```diff
@@ -78,21 +57,23 @@ describe('git operations on a repo', function() {
...
-      done(new Error('Expected an error.'));
-    }, function(err) {
-      var expectedMsg = 'error: pathspec \'' +
-                        'non-existent-branch' +
-                        '\' did not match any file(s) known to git.\n';
-      assert.notEqual(err.message.indexOf(expectedMsg), -1);
-      done();
+  it('should throw an error when checking out a non existent branch',
+    function(done) {
+      git.prepareRepo(remoteUrl,
```
Comment (shinnn) [inline test.js]:
```js
git.prepareRepo(remoteUrl, 'origin', '.publish')
```

File: test.js [hunk 29]
```diff
@@ -231,91 +194,125 @@ describe('gulp-gh-pages', function() {
...
-    .on('end', function() {
-      done(new Error('Expected an error.'));
-    })
-    .end(tmpFile);
-  });
+  it('should emit an error when the user has no permission to push the repo',
+    function(done) {
+      ghPages({remoteUrl: 'git@github.com:bot/move.git'})
+      .on('error', function(err) {
+        assert(err);
```
Comment (shinnn) [inline test.js]: Previously I checked the error message to ensure this pluigin emits a permission-related error in this case. https://github.com/shinnn/gulp-gh-pages/blob/473d4ec02fe52e3aae9c78de30bc4bc5e2b1430e/test.js#L286

Why do you think it's not needed?

**Comment by shinnn** [comment]
> I'm not willing to spend more time on this -- I already spent way too much as it is.

I'm also spending a lot of time to check unrelated code changes. That is why I proposed to create separate PR for easier review.

> I had hoped you'd be grateful to get a big improvement to your module, for free.

This is the pushiest comment that I've ever seen in my repositories :)

File: test.js (hunk 29 above)
Comment (Klortho) [inline test.js]: It was failing for me when running `mocha` locally. The error message I get is just "Command failed: /bin/sh -c git push --set-upstream origin gh-pages". Maybe something different in my environment? I don't know, but it seems too brittle to test for a specific human-readable error message in a third-party library.

**Comment by Klortho** [comment]
Sorry about the README.md whitespace changes. I had set my editor to "remove trailing whitespace", and forgot about how MD uses those. To tell you the truth, that's my least favorite feature of MD, and I never use it. I always insert `<br/>` tags if I absolutely need them -- most md processors understand those.

**Comment by Klortho** [comment]
> This is the pushiest comment that I've ever seen in my projects.

Well, I was frustrated, and it seemed like you were being pretty pushy too -- asking me to rearrange all my changes into different commits is a lot of work, and I don't see the benefit, and you never even said  "thanks for the interest".  

Anyway, as you can see, I went ahead and did everything you requested. I hope you pull it -- it's a very useful module, btw.

**Comment by shinnn** [comment]
> asking me to rearrange all my changes into different commits is a lot of work

Yes, a lot of work. So the author will end up doing a lot of work if the contributor didn't.

=====

I think it's very meaningful for you to try to contribute [ESLint](https://github.com/eslint/eslint) project. If you behave like you're doing here in that project, you'll get much more frustrated :)

ref.

* https://github.com/eslint/eslint/pull/4489
* https://twitter.com/robinberjon/status/667845758801047552
* https://twitter.com/sindresorhus/status/667996424865378309

**Comment by shinnn** [comment]
> and you never even said "thanks for the interest".

Sorry about that. I noticed this change is really helpful only after deeper insight.

No need to care about the rest of my comments https://github.com/shinnn/gulp-gh-pages/pull/83#discussion_r50623117 https://github.com/shinnn/gulp-gh-pages/pull/83#discussion_r50623082. I'll fix them by myself. Thanks for your great contribution.

File: test.js (hunk 15 above)
Comment (Klortho) [inline test.js]: No.

File: test.js (hunk 16 above)
Comment (Klortho) [inline test.js]: No

File: test.js (hunk 17 above)
Comment (Klortho) [inline test.js]: No.

File: test.js (hunk 18 above)
Comment (Klortho) [inline test.js]: No.

File: test.js (hunk 19 above)
Comment (Klortho) [inline test.js]: No.

File: test.js (hunk 20 above)
Comment (Klortho) [inline test.js]: No.

File: test.js (hunk 21 above)
Comment (Klortho) [inline test.js]: No.

File: test.js (hunk 22 above)
Comment (Klortho) [inline test.js]: No.

File: test.js (hunk 23 above)
Comment (Klortho) [inline test.js]: No.

File: test.js (hunk 24 above)
Comment (Klortho) [inline test.js]: No.

File: test.js (hunk 25 above)
Comment (Klortho) [inline test.js]: No.

File: test.js (hunk 26 above)
Comment (Klortho) [inline test.js]: No.

File: test.js (hunk 27 above)
Comment (Klortho) [inline test.js]: done

File: test.js (hunk 28 above)
Comment (Klortho) [inline test.js]: Sure, done.

**Comment by Klortho** [comment]
You're welcome, thanks for the compliment! I am not a JS expert, and learned a lot in this exercise. Thanks for your patience. I respect how strict you are with style, but would like to urge you to try to keep your line lengths shorter -- that's why I said "no" to the requests above. But, of course, it's up to you, if you want to do them yourself.
Cheers!

**Comment by Klortho** [comment]
Thanks for the links -- that was fun reading!

**Comment by Klortho** [comment]
Hi! What's the status on this? You made me jump through a lot of hoops -- I hope it wasn't for nothing!

**Comment by Klortho** [comment]
HI! Pinging you again on this. I'd rather not publish a fork on npm, but this just came up again because we have a dependency on my github fork.

After your last comment above, I had the impression that you were going to merge it. Let me know if you still have some reservation.

**Comment by eternalsky** [comment]
same problem as @Klortho ,what's the status now?

**Comment by shinnn** [comment]
Closing because there's no fix for https://github.com/shinnn/gulp-gh-pages/pull/83#r50623117 https://github.com/shinnn/gulp-gh-pages/pull/83#r50623127 https://github.com/shinnn/gulp-gh-pages/pull/83#r50623129 https://github.com/shinnn/gulp-gh-pages/pull/83#r50623131 https://github.com/shinnn/gulp-gh-pages/pull/83#r50623132 https://github.com/shinnn/gulp-gh-pages/pull/83#r50623134 https://github.com/shinnn/gulp-gh-pages/pull/83#r50623135 https://github.com/shinnn/gulp-gh-pages/pull/83#r50623137 https://github.com/shinnn/gulp-gh-pages/pull/83#r50623138 https://github.com/shinnn/gulp-gh-pages/pull/83#r50623139 https://github.com/shinnn/gulp-gh-pages/pull/83#r50623142 https://github.com/shinnn/gulp-gh-pages/pull/83#r50623154 about 2 years.

**Pull request closed without merging**